android/app/src/main/java/org/uwaterloo/subletr/pages/home/list/HomeListChildViewModel.kt

```diff
@@ -0,0 +1,196 @@
+					navHostController.navigate(
+						route = NavigationDestination.LOGIN.rootNavPath,
+						navOptions = navOptions {
+							popUpTo(navHostController.graph.id)
+						},
+					)
+				}
+					.getOrDefault(
+						GetListingsResponse(
+							listings = emptyList(),
+							liked = emptySet(),
+							pages = 0,
+						)
+					)
+			)
+		}
+		.onErrorResumeWith(Observable.never())
+		.subscribeOn(Schedulers.io())
+
+	private val imagesStream: Observable<List<Bitmap?>> = listingsStream
+		.map {
+			runBlocking {
+				it.listingsResponse.listings
+					.map { l: ListingSummary ->
+						async {
+							runCatching {
+								listingsApi.listingsImagesGet(l.imgIds.first())
+							}.getOrNull()
+						}
+					}.awaitAll()
+			}
+		}
+		.subscribeOn(Schedulers.io())
+		.observeOn(Schedulers.computation())
+		.map { base64ImageList: List<String?> ->
+			base64ImageList.map { it?.base64ToBitmap() }
+		}
+		.observeOn(Schedulers.io())
+		.onErrorResumeWith(Observable.never())
+
+	private val listingItemsStream: Observable<HomeListUiState.ListingItemsModel> = Observable.combineLatest(
+		listingsStream,
+		imagesStream,
+		::Pair
+	).map {
+		HomeListUiState.ListingItemsModel(
+			listings = it.first.listingParams.listingPagingParams.previousListingItemsModel.listings +
+				it.first.listingsResponse.listings,
+			likedListings = it.first.listingParams.listingPagingParams.previousListingItemsModel.likedListings +
+				it.first.listingsResponse.liked,
+			listingsImages = it.first.listingParams.listingPagingParams.previousListingItemsModel.listingsImages +
+				it.second,
+		)
+	}
+
+	// To ensure that view does not force an api call on every subscription
+	private val internalUiStateStream: Observable<Unit> = Observable.combineLatest(
+		locationRangeFilterStream,
+		priceRangeFilterStream,
+		roomRangeFilterStream,
+		listingItemsStream,
+		infoTextStringIdStream,
+	) { locationRange: LocationRange,
+		priceRange: PriceRange,
+		roomRange: RoomRange,
+		listings: HomeListUiState.ListingItemsModel,
+		infoTextStringId: Optional<Int>
+		->
+		uiStateStream.onNext(HomeListUiState.Loaded(
+			locationRange = locationRange,
+			priceRange = priceRange,
+			roomRange = roomRange,
+			listingItems = listings,
+			infoTextStringId = infoTextStringId.getOrNull()
+		))
+	}
+
+	override val uiStateStream: BehaviorSubject<HomeListUiState> =
+		BehaviorSubject.createDefault(HomeListUiState.Loading)
+
+	init {
+		internalUiStateStream.safeSubscribe()
+	}
+
+	data class GetListingParams(
+		val locationRange: LocationRange,
+		val priceRange: PriceRange,
+		val roomRange: RoomRange,
+		val listingPagingParams: ListingPagingParams,
+	)
+
+	data class ListingPagingParams(
+		val previousListingItemsModel: HomeListUiState.ListingItemsModel,
+		val pageNumber: Int,
+	)
+
+	data class ListingParamsAndResponse(
+		val listingParams: GetListingParams,
+		val listingsResponse: GetListingsResponse,
+	)
+
+	companion object {
+		const val LISTING_PAGE_SIZE = 5
+	}
```

Comment:
We can do this in another PR, but this would be good to dump in a constants file.

Reply:
Maybe, it is a constant specific to the `HomeListChildViewModel` so it might be fine there....

Reply:
Yeah that could work, I was thinking more there's only going to be so many constants so might as well dump them all in one file. Whatever you think is best, save for another PR though.